src/cookbooks/clients_last_seen_bits.md

```diff
@@ -0,0 +1,553 @@
+  metric_date
+WHERE
+  submission_date = '2020-01-28'
+```
+
+### Retention using activity date
+
+GUD's canonical retention definitions are all based on ping submission dates rather
+than logical activity dates taken from client-provided timestamps, but there is
+interest in using client timestamps particular for `n`-day retention calculations
+for mobile products.
+
+Let's consider Firefox Preview which sends telemetry via Glean. The
+`org_mozilla_fenix.baseline_clients_last_seen` table includes two bit patterns
+that encode client timestamps: `days_seen_session_start_bits` and
+`days_seen_session_end_bits`. This table is still populated once per day based
+on pings received over the previous day, but some of those pings will reflect
+sessions that started on previous days. This introduces some new complexity
+into retention calculations because we'll always be underestimating client
+counts if we have our retention window end on `submission_date`.
+
+When using activity date, it may be desirable to build in a few days of buffer
+to ensure we are considering late-arriving pings. For example, if we wanted
+to calculate 3-day retention but allow 2 days of cushion for late-arriving
+pings, we would need to use an offset of 5 days from `submission_date`:
+
+```
+DECLARE n, cushion_days, offset_to_day_0 INT64;
+SET n = 3;
+SET cushion_days = 2;
+SET offset_to_day_0 = -n - cushion_days;
+
+WITH base AS (
+  SELECT
+    *,
+    udf.bits28_range(days_seen_session_start_bits, offset_to_day_0 + 0, 1) AS seen_on_day_0,
+    udf.bits28_range(days_seen_session_start_bits, offset_to_day_0 + 1, n - 1) AS seen_after_day_0
+  FROM
+    org_mozilla_fenix.baseline_clients_last_seen )
+SELECT
+  DATE_SUB(submission_date, INTERVAL offset_to_day_0 DAYS) AS metric_date,
+  SAFE_DIVIDE(
+    COUNTIF(seen_on_day_0 AND seen_after_day_0),
+    COUNTIF(seen_on_day_0)
+  ) AS retention_n_day
+FROM
+  base
+GROUP BY
+  metric_date
+WHERE
+  submission_date = '2020-01-28'
+```
+
+## UDF Reference
+
+### `bits28_to_string`
+
+Convert an INT64 field into a 28 character string representing the individual bits.
+
+```
+bits28_to_string(bits INT64)
+
+SELECT udf.bits28_to_string(18)
+>> 0000000000000000000000010010
+```
+
+### `bits28_from_string`
+
+Convert a string representing individual bits into an INT64.
+
+```
+bits28_from_string(bits STRING)
+
+SELECT udf.bits28_from_string('10010')
+>> 18
+```
+
+### `bits28_to_dates`
+
+Convert a bit pattern into an array of the dates is represents.
+
+```
+bits28_to_dates(bits INT64, end_date DATE)
+
+SELECT udf.bits28_to_dates(18, '2020-01-28')
+>> ['2020-01-24', '2020-01-27']
+```
+
+### `bits28_days_since_seen`
+
+Return the position of the rightmost set bit in an INT64 bit pattern.
+
+```
+bits28_days_since_seen(bits INT64)
+
+SELECT bits28_days_since_seen(18)
+>> 1
+```
+
+
+### `bits28_range`
+
+Return an INT64 representing a range of bits from a source bit pattern.
+
+```
+bits28_range(bits INT64, offset INT64, n_bits INT64)
```

Comment:
These methods seem to aim to remove the need for users to do bitwise operations explicitly. This is probably sensible: the learning curve is steep (though also short). Though these methods are harder to debug: I can't just print out the string representation of the mask to check that I'm doing what I expected.

It took me a lot of thinking to work out whether `offset` is with respect to the leftmost or rightmost bit, and the direction in which `n_bits` counts.

Given the rest of the docs, I would expect this to be backwards-looking: I would expect `n_bits` to count leftwards and an `offset` of `0` to represent the inclusion of the rightmost bit.

But from the examples it _seems_ that `n_bits` counts rightwards? So this function is most useful for the forwards-looking usecase? Do we expect the forwards-looking usecase to be more common? Should we have separate UDFs for forwards and backwards looking usecases?

As someone most familiar with Python, I (selfishly) want to think about this stuff in terms of slicing a flipped list, with arguments `start` and `stop`. So these examples would read `udf.bits28_range(18, 0, 6)`, `udf.bits28_range(18, 4, 6)` and `udf.bits28_range(18, 0, 4)` respectively.

Reply:
> But from the examples it seems that n_bits counts rightwards? So this function is most useful for the forwards-looking usecase? Do we expect the forwards-looking usecase to be more common? Should we have separate UDFs for forwards and backwards looking usecases?

You are correct on all accounts. For backwards-looking use cases, `bits28_days_since_seen` is the only UDF needed. `bits28_range`, `bits28_active_in_range`, and `bits28_retention` are all intended for forward-looking windows, so they expect a negative offset to determine how many bits left to move before it starts reading, then `n_bits` measures how many bits to the right to move.

Reply:
> As someone most familiar with Python, I (selfishly) want to think about this stuff in terms of slicing a flipped list, with arguments start and stop

What you describe is definitely more straight-forward in terms of how the parameters relate to the bit pattern. But that interface doesn't seem helpful for thinking about forward-looking windows where I think the right thing to do is to define a starting bit and then count rightward.

Does that sound reasonable? I can certainly add more to the documentation of this function to clarify some of the points you found surprising.

Reply:
> For backwards-looking use cases, bits28_days_since_seen is the only UDF needed.

Not for my backwards-looking use cases! :sweat_smile:

Since the text above already discusses both forward and backward windows at length, it probably makes sense to specify which of these UDFs are counting forward and which are counting backward. At least here in the docs; optionally also in the naming of the functions, e.g. `bits28_forward_range()`.

Reply:
> Does that sound reasonable?

Yes! As long as it's clear that they're for forward looking windows

Reply:
Well, we could modify this function to allow negative numbers for `n_bits`. Negative `n_bits` would mean that we count left from the start offset (backward-looking windows) while positive `n_bits` would mean counting right (forward-looking windows).

Reply:
For range specification/slicing, `(start, stop)` is an established format, as is `(start, length)` for positive lengths. I don't think I've seen `(start, length)` overloaded to become `(stop, negative_length)` when `length < 0`.

I would be a bit too scared to use it and will probably stick with masks (I'm OK with that - past the learning curve now) - I would worry about messing up the arithmetic and doing off-by-one miscalculations that take me to the 29th bit, yielding plausible but wrong results.

Reply:
> I would worry about messing up the arithmetic and doing off-by-one miscalculations that take me to the 29th bit

We already have validation logic in this UDF such that we'll raise an error `'Reading %i bits from starting_offset %i exceeds end of bit pattern in call bits28_range(%i, %i, %i)'`. We would add additional checks if we enabled negative `n_bits` to at least catch the "29th bit" case.

> For range specification/slicing, (start, stop) is an established format, as is (start, length) for positive lengths. I don't think I've seen (start, length) overloaded to become (stop, negative_length) when length < 0.

I feel you here. I'm not sure I've ever seen an API that's quite analogous to what we have here. To be logically consistent for forward and backward-looking cases, I think we need to be consistent in thinking of offsets in terms of higher values as we move to the right. But maybe you're right that it would be easier to reason about if we had separate `bits28_range_forward` and `bits28_range_backward` or `bits28_range_right` and `bits28_range_left`.